lib/ide/src/controller/text.rs

```diff
@@ -0,0 +1,118 @@
+//! module file and plain text file. In case of luna module idmap and metadata are hidden for the
+//! user.
+
+use crate::prelude::*;
+
+use crate::controller::*;
+
+use failure::_core::fmt::{Formatter, Error};
+use flo_stream::{Publisher, Subscriber};
+use flo_stream::MessagePublisher;
+use file_manager_client as fmc;
+use json_rpc::error::RpcError;
+use shapely::shared;
+
+
+
+// ====================
+// === Notification ===
+// ====================
+
+/// A notification from TextController.
+#[derive(Clone,Debug)]
+pub enum Notification {
+    /// File contents needs to be set to the following due to synchronization with external state.
+    SetNewContent(String),
+}
+
+
+
+// =======================
+// === Text Controller ===
+// =======================
+
+shared! { ControllerHandle
+
+    /// Data stored by the text controller.
+    pub struct State {
+        /// A path to file the controller is handling.
+        file_path : fmc::Path,
+        /// A module handle in case the TextController is managing Luna module file.
+        module: Option<module::ControllerHandle>,
+        /// Sink where we put events to be consumed by the view.
+        notification_publisher: Publisher<Notification>,
+        /// File manager handle, used to obtain file information in case of plain text file.
+        file_manager: fmc::ClientHandle,
+    }
+
+    impl {
+        /// Create controller managing plain text file.
+        pub fn new(path:fmc::Path, file_manager:fmc::ClientHandle) -> Self {
+            Self {file_manager,
+                file_path              : path,
+                module                 : None,
+                notification_publisher : Publisher::new(10),
+            }
+        }
+
+        /// Get subscriber receiving controller's notifications.
+        pub fn subscribe(&mut self) -> Subscriber<Notification> {
+            self.notification_publisher.subscribe()
+        }
+
+        /// Get clone of file path handled by this controller.
+        pub fn file_path_clone(&self) -> fmc::Path {
+            self.file_path.clone()
+        }
+    }
+}
+
+impl ControllerHandle {
+    /// Create controller managing Luna module file.
+    pub fn new_for_module(module:module::ControllerHandle, file_manager:fmc::ClientHandle) -> Self {
+        let file_path = module.location_as_path();
+        let mut state = State::new(file_path, file_manager);
+        state.module = Some(module);
+        Self { rc:Rc::new(RefCell::new(state)) }
+    }
+
+    /// Read file's content.
+    pub fn read_content(&self) -> impl Future<Output=Result<String,RpcError>> {
+        let (mut file_manager,path) = self.with_borrowed(|state| {
+            (state.file_manager.clone_ref(), state.file_path.clone())
+        });
+        file_manager.read(path)
+    }
+
+    /// Store the given content to file.
+    pub fn store_content(&self, content:String) -> impl Future<Output=Result<(),RpcError>> {
```

Comment:
@mwu-tow I don't get your comment. This function is about `storing content`. I dont see any part about metadata removal.

Reply:
Now it is fine, as it has separate branch for module.

Previously, it would just flush the text to the module — and that would truncate the metadata, as they are stored separately and known about by module controller only.